### PR TITLE
doc, test: Compilation for 64-bit Windows with msys2

### PR DESCRIPTION
**Pull request by rnapoles**
* update steps for build Bitcoin Core with msys2
* fix test/util/rpcauth-test.py and test/util/test_runner.py to run in msys2
* added test/util/path_helper.py to handle msys2 paths in python

### CODE REVIEW

File: doc/build-windows.md [hunk 1]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+  pacman -Sy
+  pacman -Su
+  pacman -S base-devel gcc vim cmake git mingw-w64-{x86_64,i686}-nsis
+  pacman -S --needed filesystem msys2-runtime bash libreadline libiconv libarchive libgpgme libcurl pacman ncurses libintl
+  pacman -S make autoconf autotools automake libtool pkg-config autotools-dev pkg-config
+  pacman -S mingw-w64-{i686,x86_64}-toolchain mingw-w64-{i686,x86_64}-pkgconf
+  pacman -S mingw-w64-{i686,x86_64}-{boost,protobuf}
+  pacman -S mingw-w64-{i686,x86_64}-crt-git
+  pacman -S mingw-w64-{i686,x86_64}-libevent
+  pacman -S mingw-w64-{i686,x86_64}-db mingw-w64-{i686,x86_64}-leveldb
```
Comment (fanquake) [inline doc/build-windows.md:111]: leveldb is part of our source. It doesn't need installing.

File: doc/build-windows.md [hunk 2]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+
+#### Build using:
+```bash
+  ./autogen.sh
+
+  ./configure \
+      --with-gui=no \
+      --without-miniupnpc \
+      --without-bdb \
+      --without-gui \
```
Comment (fanquake) [inline doc/build-windows.md:128]: `--with-gui=no` and `--without-gui` do the same thing.

File: doc/build-windows.md [hunk 3]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+      --with-qrencode
```
Comment (fanquake) [inline doc/build-windows.md:129]: Why are you passing `--with-qrencode` (gui component) if you've explicitly disabled the GUI?

File: doc/build-windows.md [hunk 4]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+
+#### Install
+```bash
+  pacman -Sy
+  pacman -Su
+  pacman -S base-devel gcc vim cmake git mingw-w64-{x86_64,i686}-nsis
+  pacman -S --needed filesystem msys2-runtime bash libreadline libiconv libarchive libgpgme libcurl pacman ncurses libintl
+  pacman -S make autoconf autotools automake libtool pkg-config autotools-dev pkg-config
+  pacman -S mingw-w64-{i686,x86_64}-toolchain mingw-w64-{i686,x86_64}-pkgconf
+  pacman -S mingw-w64-{i686,x86_64}-{boost,protobuf}
```
[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md]: included in updated doc/build-windows.md

File: doc/build-windows.md [hunk 5]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+  pacman -S mingw-w64-{i686,x86_64}-crt-git
+  pacman -S mingw-w64-{i686,x86_64}-libevent
+  pacman -S mingw-w64-{i686,x86_64}-db mingw-w64-{i686,x86_64}-leveldb
+  pacman -S mingw-w64-{i686,x86_64}-miniupnpc
```
[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md:112]: included in updated doc/build-windows.md

File: doc/build-windows.md (hunk 1 above)
Comment (rnapoles) [reply doc/build-windows.md]: included in updated doc/build-windows.md

File: doc/build-windows.md (hunk 2 above)
Comment (rnapoles) [reply doc/build-windows.md]: included in updated doc/build-windows.md

File: doc/build-windows.md (hunk 3 above)
Comment (rnapoles) [reply doc/build-windows.md]: included in updated doc/build-windows.md

File: doc/build-windows.md [hunk 6]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+      --without-bdb \
+      --without-gui \
+      --with-qrencode
+
+  make -J
+
+  make check
+```
+
+#### Credits
```
[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md]: rename to  "References" doc/build-windows.md, What do you think about this?

File: doc/build-windows.md [hunk 7]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+#### Execute:
+```bash
+  C:\msys64\msys2.exe
+```
+
+#### Install
+```bash
+  pacman -Sy
+  pacman -Su
+  pacman -S base-devel gcc vim cmake git mingw-w64-{x86_64,i686}-nsis
```
[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md]: remove in updated doc/build-windows.md

File: doc/build-windows.md [hunk 8]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+  ./autogen.sh
+
+  ./configure \
+      --with-gui=no \
+      --without-miniupnpc \
+      --without-bdb \
+      --without-gui \
+      --with-qrencode
+
+  make -J
```
[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md]: corrected in updated doc/build-windows.md

File: doc/build-windows.md [hunk 9]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+```
+
+#### Install
+```bash
+  pacman -Sy
+  pacman -Su
+  pacman -S base-devel gcc vim cmake git mingw-w64-{x86_64,i686}-nsis
+  pacman -S --needed filesystem msys2-runtime bash libreadline libiconv libarchive libgpgme libcurl pacman ncurses libintl
+  pacman -S make autoconf autotools automake libtool pkg-config autotools-dev pkg-config
+  pacman -S mingw-w64-{i686,x86_64}-toolchain mingw-w64-{i686,x86_64}-pkgconf
```
[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md:107]: corrected in updated doc/build-windows.md

File: doc/build-windows.md [hunk 10]
```diff
@@ -87,6 +87,56 @@ Build using:
...
+```bash
+  C:\msys64\msys2.exe
+```
+
+#### Install
+```bash
+  pacman -Sy
+  pacman -Su
+  pacman -S base-devel gcc vim cmake git mingw-w64-{x86_64,i686}-nsis
+  pacman -S --needed filesystem msys2-runtime bash libreadline libiconv libarchive libgpgme libcurl pacman ncurses libintl
```
[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md]: corrected in updated doc/build-windows.md

File: doc/build-windows.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply doc/build-windows.md]: > What do you think about this?

I still don't think there's any need for a references, or credits section. Let alone one that links to a repository that hasn't been updated in 7 years.

[the remark this answers is not in the document]
Comment (rnapoles) [reply doc/build-windows.md]: I understand, removed outdated references in new commit

**Comment by fanquake** [comment]
@rnapoles note that you'll need to squash / cleanup your commits (no need to open a new PR).

@sipsorcery any thoughts on this?

**Comment by sipsorcery** [comment]
@rnapoles I'd recommend splitting this PR up. Some of the changes seem to be pure formatting, some are related to the msys2 change (which is all that is expected given the PR title) and some seem to be changes to the test scripts. It's a lot easier to get reviewers and ACK's on PRs if they are single purpose.

In regards to the msys2 change my personal opinion would be that it's better to stick with mingw because that's what the Bitcoin Core guix builds are done with.

**Comment by MarcoFalke** [comment]
I think the plan is to move from mingw to clang windows cross builds

**Comment by fanquake** [comment]
> In regards to the msys2 change my personal opinion would be that it's better to stick with mingw because that's what the Bitcoin Core guix builds are done with.

Closing for now.

**Pull request closed without merging**